File: impl/go/candidate.go
package twill

import "strings"

// ValueKind distinguishes named and arbitrary values and modifiers.
type ValueKind int

const (
	// ValueNamed is a named value such as `red-500`.
	ValueNamed ValueKind = iota
	// ValueArbitrary is a decoded arbitrary value.
	ValueArbitrary
)

// CandidateValue is the value segment of a functional candidate (SPEC §4.1.4).
type CandidateValue struct {
	Kind  ValueKind
	Value string
	// Fraction is set for named values when a slash segment could be a
	// fraction, for example `1/2`.
	Fraction *string
	// DataType is an explicit type hint of an arbitrary value.
	DataType *string
}

// Modifier is a candidate or variant modifier (SPEC §4.1.5).
type Modifier struct {
	Kind  ValueKind
	Value string
}

// VariantKind is the kind of a parsed variant.
type VariantKind int

const (
	// VariantStatic is a static variant.
	VariantStatic VariantKind = iota
	// VariantFunctional is a functional variant.
	VariantFunctional
	// VariantCompound is a compound variant.
	VariantCompound
	// VariantArbitrary is an arbitrary variant.
	VariantArbitrary
)

// VariantValue is the value of a functional variant.
type VariantValue struct {
	Kind  ValueKind
	Value string
}

// Variant is a parsed variant (SPEC §4.1.6).
type Variant struct {
	Kind     VariantKind
	Root     string
	Value    *VariantValue
	Modifier *Modifier
	// Inner is the inner variant of a compound variant.
	Inner *Variant
	// Selector and Relative belong to arbitrary variants.
	Selector string
	Relative bool
}

// CandidateKind is the kind of a parsed candidate.
type CandidateKind int

const (
	// CandidateStatic is a static utility.
	CandidateStatic CandidateKind = iota
	// CandidateFunctional is a functional utility.
	CandidateFunctional
	// CandidateArbitrary is an arbitrary property.
	CandidateArbitrary
)

// Candidate is a parsed class name (SPEC §4.1.3).
type Candidate struct {
	Kind CandidateKind
	// Raw is the original class name including variants and the important marker.
	Raw string
	// Variants are in application order: the rightmost in the source is first.
	Variants  []*Variant
	Important bool
	// Root is set for static and functional candidates.
	Root string
	// Value and Modifier are set for functional candidates.
	Value    *CandidateValue
	Modifier *Modifier
	// Property and ArbitraryValue are set for arbitrary candidates.
	Property       string
	ArbitraryValue string
}

// UtilityKind is the kind of a utility definition.
type UtilityKind int

const (
	// UtilityStatic is a static utility.
	UtilityStatic UtilityKind = iota
	// UtilityFunctional is a functional utility.
	UtilityFunctional
)

// ParserContext is the part of the design system the parsers consult.
type ParserContext interface {
	ThemePrefix() string
	HasUtility(name string, kind UtilityKind) bool
	HasVariant(name string) bool
	VariantKindOf(name string) (VariantKind, bool)
	CompoundsWith(parent string, child *Variant) bool
	// ParseVariant is memoized; it returns the same object for the same input.
	ParseVariant(input string) *Variant
}

// RootMatch is a split of an input into a root and an optional value.
type RootMatch struct {
	Root  string
	Value *string
}

// FindRoots yields every (root, value) split whose root exists (SPEC §8.2.2).
func FindRoots(input string, exists func(string) bool) []RootMatch {
	var result []RootMatch
	if exists(input) {
		result = append(result, RootMatch{Root: input})
	}
	idx := strings.LastIndex(input, "-")
	if idx != -1 {
		for {
			root := input[:idx]
			if exists(root) {
				value := input[idx+1:]
				if value == "" {
					break
				}
				if root == "@" {
					break
				}
				result = append(result, RootMatch{Root: root, Value: strPtr(value)})
			}
			if idx == 0 {
				break
			}
			idx = strings.LastIndex(input[:idx], "-")
			if idx <= 0 {
				break
			}
		}
	}
	if strings.HasPrefix(input, "@") && exists("@") {
		result = append(result, RootMatch{Root: "@", Value: strPtr(input[1:])})
	}
	return result
}

// ParseModifier parses a modifier segment (SPEC §8.2.1).
func ParseModifier(modifier string) *Modifier {
	if strings.HasPrefix(modifier, "[") && strings.HasSuffix(modifier, "]") {
		value := DecodeArbitraryValue(modifier[1 : len(modifier)-1])
		if !IsValidArbitrary(value) || strings.TrimSpace(value) == "" {
			return nil
		}
		return &Modifier{Kind: ValueArbitrary, Value: value}
	}
	if strings.HasPrefix(modifier, "(") && strings.HasSuffix(modifier, ")") {
		value := DecodeArbitraryValue(modifier[1 : len(modifier)-1])
		if !strings.HasPrefix(value, "--") || !IsValidArbitrary(value) {
			return nil
		}
		return &Modifier{Kind: ValueArbitrary, Value: "var(" + value + ")"}
	}
	if IsNamedValue(modifier) {
		return &Modifier{Kind: ValueNamed, Value: modifier}
	}
	return nil
}

// ParseCandidate returns every interpretation of a raw class name (SPEC §8.2).
func ParseCandidate(input string, ds ParserContext) []*Candidate {
	var results []*Candidate

	rawVariants := Segment(input, ':')
	if prefix := ds.ThemePrefix(); prefix != "" {
		if len(rawVariants) == 1 || rawVariants[0] != prefix {
			return nil
		}
		rawVariants = rawVariants[1:]
	}
	base := rawVariants[len(rawVariants)-1]
	rawVariants = rawVariants[:len(rawVariants)-1]

	variants := []*Variant{}
	for i := len(rawVariants) - 1; i >= 0; i-- {
		v := ds.ParseVariant(rawVariants[i])
		if v == nil {
			return nil
		}
		variants = append(variants, v)
	}

	important := false
	if strings.HasSuffix(base, "!") {
		important = true
		base = base[:len(base)-1]
	} else if strings.HasPrefix(base, "!") {
		important = true
		base = base[1:]
	}

	if ds.HasUtility(base, UtilityStatic) && !strings.Contains(base, "[") {
		results = append(results, &Candidate{Kind: CandidateStatic, Root: base, Variants: variants, Important: important, Raw: input})
	}

	parts := Segment(base, '/')
	if len(parts) > 2 {
		return results
	}
	baseWithoutModifier := parts[0]
	var modifierSegment *string
	if len(parts) == 2 {
		modifierSegment = strPtr(parts[1])
	}
	var modifier *Modifier
	if modifierSegment != nil {
		modifier = ParseModifier(*modifierSegment)
		if modifier == nil {
			return results
		}
	}

	if strings.HasPrefix(baseWithoutModifier, "[") {
		if !strings.HasSuffix(baseWithoutModifier, "]") || len(baseWithoutModifier) < 2 {
			return results
		}
		second := baseWithoutModifier[1]
		if !(second == '-' || (second >= 'a' && second <= 'z')) {
			return results
		}
		inner := baseWithoutModifier[1 : len(baseWithoutModifier)-1]
		colon := strings.Index(inner, ":")
		if colon == -1 || colon == 0 || colon == len(inner)-1 {
			return results
		}
		value := DecodeArbitraryValue(inner[colon+1:])
		if !IsValidArbitrary(value) {
			return results
		}
		return append(results, &Candidate{
			Kind: CandidateArbitrary, Property: inner[:colon], ArbitraryValue: value,
			Modifier: modifier, Variants: variants, Important: important, Raw: input,
		})
	}

	var roots []RootMatch
	switch {
	case strings.HasSuffix(baseWithoutModifier, "]"):
		idx := strings.Index(baseWithoutModifier, "-[")
		if idx == -1 {
			return results
		}
		root := baseWithoutModifier[:idx]
		if !ds.HasUtility(root, UtilityFunctional) {
			return results
		}
		roots = []RootMatch{{Root: root, Value: strPtr(baseWithoutModifier[idx+1:])}}
	case strings.HasSuffix(baseWithoutModifier, ")"):
		idx := strings.Index(baseWithoutModifier, "-(")
		if idx == -1 {
			return results
		}
		root := baseWithoutModifier[:idx]
		if !ds.HasUtility(root, UtilityFunctional) {
			return results
		}
		inner := baseWithoutModifier[idx+2 : len(baseWithoutModifier)-1]
		innerParts := Segment(inner, ':')
		dataType := ""
		value := inner
		if len(innerParts) == 2 {
			dataType, value = innerParts[0], innerParts[1]
		} else if len(innerParts) > 2 {
			return results
		}
		if !strings.HasPrefix(value, "--") || !IsValidArbitrary(value) {
			return results
		}
		rewritten := "[var(" + value + ")]"
		if dataType != "" {
			rewritten = "[" + dataType + ":var(" + value + ")]"
		}
		roots = []RootMatch{{Root: root, Value: strPtr(rewritten)}}
	default:
		roots = FindRoots(baseWithoutModifier, func(r string) bool { return ds.HasUtility(r, UtilityFunctional) })
	}

	for _, match := range roots {
		candidate := &Candidate{Kind: CandidateFunctional, Root: match.Root, Modifier: modifier, Variants: variants, Important: important, Raw: input}
		if match.Value == nil {
			results = append(results, candidate)
			continue
		}
		value := *match.Value
		if bracket := strings.Index(value, "["); bracket != -1 {
			if !strings.HasSuffix(value, "]") {
				return results
			}
			decoded := DecodeArbitraryValue(value[bracket+1 : len(value)-1])
			if !IsValidArbitrary(decoded) {
				continue
			}
			var dataType *string
			arbitraryValue := decoded
			i := 0
			for i < len(decoded) {
				c := decoded[i]
				if c == '-' || (c >= 'a' && c <= 'z') {
					i++
				} else {
					break
				}
			}
			if i < len(decoded) && decoded[i] == ':' {
				hint := decoded[:i]
				if hint == "" {
					continue
				}
				dataType = strPtr(hint)
				arbitraryValue = decoded[i+1:]
			}
			if strings.TrimSpace(arbitraryValue) == "" {
				continue
			}
			candidate.Value = &CandidateValue{Kind: ValueArbitrary, Value: arbitraryValue, DataType: dataType}
		} else {
			var fraction *string
			if modifierSegment != nil && modifier != nil && modifier.Kind == ValueNamed {
				fraction = strPtr(value + "/" + *modifierSegment)
			}
			if !IsNamedValue(value) {
				continue
			}
			candidate.Value = &CandidateValue{Kind: ValueNamed, Value: value, Fraction: fraction}
		}
		results = append(results, candidate)
	}
	return results
}

// ParseVariant parses a variant segment (SPEC §8.3).
func ParseVariant(input string, ds ParserContext) *Variant {
	if strings.HasPrefix(input, "[") && strings.HasSuffix(input, "]") {
		if len(input) > 1 && input[1] == '@' && strings.Contains(input, "&") {
			return nil
		}
		selector := DecodeArbitraryValue(input[1 : len(input)-1])
		if !IsValidArbitrary(selector) || strings.TrimSpace(selector) == "" {
			return nil
		}
		first := selector[0]
		relative := first == '>' || first == '+' || first == '~'
		result := selector
		if !relative && first != '@' && !strings.Contains(selector, "&") {
			result = "&:is(" + selector + ")"
		}
		return &Variant{Kind: VariantArbitrary, Selector: result, Relative: relative}
	}

	parts := Segment(input, '/')
	if len(parts) > 2 {
		return nil
	}
	name := parts[0]
	var modifierText *string
	if len(parts) == 2 {
		modifierText = strPtr(parts[1])
	}

	for _, match := range FindRoots(name, ds.HasVariant) {
		kind, ok := ds.VariantKindOf(match.Root)
		if !ok {
			continue
		}
		switch kind {
		case VariantStatic:
			if match.Value != nil || modifierText != nil {
				return nil
			}
			return &Variant{Kind: VariantStatic, Root: match.Root}
		case VariantFunctional:
			var modifier *Modifier
			if modifierText != nil {
				modifier = ParseModifier(*modifierText)
				if modifier == nil {
					return nil
				}
			}
			if match.Value == nil {
				return &Variant{Kind: VariantFunctional, Root: match.Root, Modifier: modifier}
			}
			value := *match.Value
			if strings.HasSuffix(value, "]") {
				if !strings.HasPrefix(value, "[") {
					continue
				}
				decoded := DecodeArbitraryValue(value[1 : len(value)-1])
				if !IsValidArbitrary(decoded) || strings.TrimSpace(decoded) == "" {
					return nil
				}
				return &Variant{Kind: VariantFunctional, Root: match.Root, Value: &VariantValue{Kind: ValueArbitrary, Value: decoded}, Modifier: modifier}
			}
			if strings.HasSuffix(value, ")") {
				if !strings.HasPrefix(value, "(") {
					continue
				}
				decoded := DecodeArbitraryValue(value[1 : len(value)-1])
				if !IsValidArbitrary(decoded) || strings.TrimSpace(decoded) == "" || !strings.HasPrefix(decoded, "--") {
					return nil
				}
				return &Variant{Kind: VariantFunctional, Root: match.Root, Value: &VariantValue{Kind: ValueArbitrary, Value: "var(" + decoded + ")"}, Modifier: modifier}
			}
			if !IsNamedValue(value) {
				continue
			}
			return &Variant{Kind: VariantFunctional, Root: match.Root, Value: &VariantValue{Kind: ValueNamed, Value: value}, Modifier: modifier}
		case VariantCompound:
			if match.Value == nil {
				return nil
			}
			innerText := *match.Value
			remaining := modifierText
			if (match.Root == "not" || match.Root == "has" || match.Root == "in") && modifierText != nil {
				innerText = innerText + "/" + *modifierText
				remaining = nil
			}
			inner := ds.ParseVariant(innerText)
			if inner == nil {
				return nil
			}
			if !ds.CompoundsWith(match.Root, inner) {
				return nil
			}
			var modifier *Modifier
			if remaining != nil {
				modifier = ParseModifier(*remaining)
				if modifier == nil {
					return nil
				}
			}
			return &Variant{Kind: VariantCompound, Root: match.Root, Modifier: modifier, Inner: inner}
		}
	}
	return nil
}

File: impl/ocaml/src/candidate.ml
(* Candidate and variant parsing (SPEC §4.1.3 through §4.1.6, §8). *)

open Utils

type value_kind = Named | Arbitrary

(* The value segment of a functional candidate. [fraction] is set for named
   values when a slash segment could be a fraction; [data_type] is an
   explicit type hint of an arbitrary value. *)
type candidate_value = { vkind : value_kind; value : string; fraction : string option; data_type : string option }

type modifier = { mkind : value_kind; mvalue : string }
type variant_kind = Static_variant | Functional_variant | Compound_variant | Arbitrary_variant
type variant_value = { vvkind : value_kind; vvalue : string }

type variant = {
  kind : variant_kind;
  root : string;
  vvalue : variant_value option;
  vmodifier : modifier option;
  inner : variant option;
  (* [selector] and [relative] belong to arbitrary variants. *)
  selector : string;
  relative : bool;
  (* The text this variant was parsed from; memoized parsers key on it. *)
  vraw : string;
}

type candidate_kind = Static_candidate | Functional_candidate | Arbitrary_candidate

type candidate = {
  ckind : candidate_kind;
  raw : string;
  (* In application order: the rightmost in the source is first. *)
  variants : variant list;
  important : bool;
  croot : string;
  cvalue : candidate_value option;
  cmodifier : modifier option;
  property : string;
  arbitrary_value : string;
}

type utility_kind = At_utility.kind = Static | Functional

(* The part of the design system the parsers consult. *)
type context = {
  theme_prefix : unit -> string;
  has_utility : string -> utility_kind -> bool;
  has_variant : string -> bool;
  variant_kind_of : string -> variant_kind option;
  compounds_with : string -> variant -> bool;
  (* Memoized: returns the same object for the same input. *)
  parse_variant : string -> variant option;
}

let make_variant ?(vvalue = None) ?(vmodifier = None) ?(inner = None) ?(selector = "") ?(relative = false) ?(vraw = "") kind root =
  { kind; root; vvalue; vmodifier; inner; selector; relative; vraw }

let named_value ?(fraction = None) value = { vkind = Named; value; fraction; data_type = None }
let arbitrary_value ?(data_type = None) value = { vkind = Arbitrary; value; fraction = None; data_type }

let make_candidate ?(variants = []) ?(important = false) ?(croot = "") ?(cvalue = None) ?(cmodifier = None) ?(property = "")
    ?(arbitrary_value = "") ckind raw =
  { ckind; raw; variants; important; croot; cvalue; cmodifier; property; arbitrary_value }

type root_match = { rroot : string; rvalue : string option }

(* Every (root, value) split whose root exists (SPEC §8.2.2). *)
let find_roots input exists =
  let result = ref [] in
  if exists input then result := [ { rroot = input; rvalue = None } ];
  (match String.rindex_opt input '-' with
  | None -> ()
  | Some idx ->
      let idx = ref idx in
      let continue = ref true in
      while !continue do
        let root = String.sub input 0 !idx in
        if exists root then begin
          let value = after input (!idx + 1) in
          if value = "" || root = "@" then continue := false
          else result := !result @ [ { rroot = root; rvalue = Some value } ]
        end;
        if !continue then
          if !idx = 0 then continue := false
          else
            match String.rindex_opt (String.sub input 0 !idx) '-' with
            | Some j when j > 0 -> idx := j
            | _ -> continue := false
      done);
  if has_prefix input "@" && exists "@" then result := !result @ [ { rroot = "@"; rvalue = Some (after input 1) } ];
  !result

(* Parses a modifier segment (SPEC §8.2.1). *)
let parse_modifier modifier =
  let n = String.length modifier in
  if has_prefix modifier "[" && has_suffix modifier "]" then begin
    let value = decode_arbitrary_value (String.sub modifier 1 (n - 2)) in
    if (not (is_valid_arbitrary value)) || String.trim value = "" then None else Some { mkind = Arbitrary; mvalue = value }
  end
  else if has_prefix modifier "(" && has_suffix modifier ")" then begin
    let value = decode_arbitrary_value (String.sub modifier 1 (n - 2)) in
    if (not (has_prefix value "--")) || not (is_valid_arbitrary value) then None
    else Some { mkind = Arbitrary; mvalue = "var(" ^ value ^ ")" }
  end
  else if is_named_value modifier then Some { mkind = Named; mvalue = modifier }
  else None

exception Done

(* Returns every interpretation of a raw class name (SPEC §8.2). *)
let parse_candidate input (ds : context) : candidate list =
  let results = ref [] in
  let raw_variants = segment input ':' in
  let prefix = ds.theme_prefix () in
  let raw_variants =
    if prefix <> "" then
      match raw_variants with [ _ ] -> raise Done | p :: rest when p = prefix -> rest | _ -> raise Done
    else raw_variants
  in
  let base = List.nth raw_variants (List.length raw_variants - 1) in
  let raw_variants = List.filteri (fun i _ -> i < List.length raw_variants - 1) raw_variants in
  let variants =
    List.map
      (fun v -> match ds.parse_variant v with Some p -> p | None -> raise Done)
      (List.rev raw_variants)
  in
  let important, base =
    if has_suffix base "!" then (true, String.sub base 0 (String.length base - 1))
    else if has_prefix base "!" then (true, after base 1)
    else (false, base)
  in
  if ds.has_utility base Static && not (String.contains base '[') then
    results := [ make_candidate ~variants ~important ~croot:base Static_candidate input ];
  let finish () = !results in
  let parts = segment base '/' in
  if List.length parts > 2 then finish ()
  else begin
    let base_without_modifier = List.hd parts in
    let modifier_segment = match parts with [ _; m ] -> Some m | _ -> None in
    let modifier = match modifier_segment with Some m -> parse_modifier m | None -> None in
    if modifier_segment <> None && modifier = None then finish ()
    else if has_prefix base_without_modifier "[" then begin
      let b = base_without_modifier in
      let n = String.length b in
      if (not (has_suffix b "]")) || n < 2 then finish ()
      else
        let second = b.[1] in
        if not (second = '-' || is_lower second) then finish ()
        else
          let inner = String.sub b 1 (n - 2) in
          match String.index_opt inner ':' with
          | None -> finish ()
          | Some colon when colon = 0 || colon = String.length inner - 1 -> finish ()
          | Some colon ->
              let value = decode_arbitrary_value (after inner (colon + 1)) in
              if not (is_valid_arbitrary value) then finish ()
              else
                !results
                @ [ make_candidate ~variants ~important ~cmodifier:modifier ~property:(String.sub inner 0 colon)
                      ~arbitrary_value:value Arbitrary_candidate input ]
    end
    else begin
      let b = base_without_modifier in
      let roots =
        if has_suffix b "]" then begin
          match index_of b "-[" with
          | -1 -> raise Done
          | idx ->
              let root = String.sub b 0 idx in
              if not (ds.has_utility root Functional) then raise Done;
              [ { rroot = root; rvalue = Some (after b (idx + 1)) } ]
        end
        else if has_suffix b ")" then begin
          match index_of b "-(" with
          | -1 -> raise Done
          | idx ->
              let root = String.sub b 0 idx in
              if not (ds.has_utility root Functional) then raise Done;
              let inner = String.sub b (idx + 2) (String.length b - idx - 3) in
              let data_type, value =
                match segment inner ':' with
                | [ v ] -> ("", v)
                | [ t; v ] -> (t, v)
                | _ -> raise Done
              in
              if (not (has_prefix value "--")) || not (is_valid_arbitrary value) then raise Done;
              let rewritten = if data_type = "" then "[var(" ^ value ^ ")]" else "[" ^ data_type ^ ":var(" ^ value ^ ")]" in
              [ { rroot = root; rvalue = Some rewritten } ]
        end
        else find_roots b (fun r -> ds.has_utility r Functional)
      in
      let rec loop = function
        | [] -> ()
        | m :: rest -> (
            let base_candidate = make_candidate ~variants ~important ~croot:m.rroot ~cmodifier:modifier Functional_candidate input in
            match m.rvalue with
            | None ->
                results := !results @ [ base_candidate ];
                loop rest
            | Some value ->
                let skip = ref false in
                let cvalue =
                  match String.index_opt value '[' with
                  | Some bracket ->
                      if not (has_suffix value "]") then raise Done;
                      let decoded = decode_arbitrary_value (String.sub value (bracket + 1) (String.length value - bracket - 2)) in
                      if not (is_valid_arbitrary decoded) then (skip := true; None)
                      else begin
                        let i = ref 0 in
                        let dn = String.length decoded in
                        while !i < dn && (decoded.[!i] = '-' || is_lower decoded.[!i]) do incr i done;
                        let data_type, arbitrary =
                          if !i < dn && decoded.[!i] = ':' then
                            let hint = String.sub decoded 0 !i in
                            if hint = "" then (skip := true; (None, decoded)) else (Some hint, after decoded (!i + 1))
                          else (None, decoded)
                        in
                        if String.trim arbitrary = "" then skip := true;
                        Some (arbitrary_value ~data_type arbitrary)
                      end
                  | None ->
                      let fraction =
                        match (modifier_segment, modifier) with
                        | Some seg, Some { mkind = Named; _ } -> Some (value ^ "/" ^ seg)
                        | _ -> None
                      in
                      if not (is_named_value value) then (skip := true; None) else Some (named_value ~fraction value)
                in
                if not !skip then results := !results @ [ { base_candidate with cvalue } ];
                loop rest)
      in
      loop roots;
      finish ()
    end
  end

let parse_candidate input ds = try parse_candidate input ds with Done -> []

(* Parses a variant segment (SPEC §8.3). *)
let parse_variant input (ds : context) : variant option =
  let n = String.length input in
  if has_prefix input "[" && has_suffix input "]" then begin
    if n > 1 && input.[1] = '@' && String.contains input '&' then None
    else
      let selector = decode_arbitrary_value (String.sub input 1 (n - 2)) in
      if (not (is_valid_arbitrary selector)) || String.trim selector = "" then None
      else
        let first = selector.[0] in
        let relative = first = '>' || first = '+' || first = '~' in
        let result = if (not relative) && first <> '@' && not (String.contains selector '&') then "&:is(" ^ selector ^ ")" else selector in
        Some (make_variant ~selector:result ~relative ~vraw:input Arbitrary_variant "")
  end
  else
    let parts = segment input '/' in
    if List.length parts > 2 then None
    else
      let name = List.hd parts in
      let modifier_text = match parts with [ _; m ] -> Some m | _ -> None in
      let rec try_roots = function
        | [] -> None
        | m :: rest -> (
            match ds.variant_kind_of m.rroot with
            | None -> try_roots rest
            | Some Static_variant -> if m.rvalue <> None || modifier_text <> None then None else Some (make_variant ~vraw:input Static_variant m.rroot)
            | Some Functional_variant -> (
                let modifier =
                  match modifier_text with
                  | Some t -> ( match parse_modifier t with Some md -> Some (Some md) | None -> None)
                  | None -> Some None
                in
                match modifier with
                | None -> None
                | Some vmodifier -> (
                    match m.rvalue with
                    | None -> Some (make_variant ~vmodifier ~vraw:input Functional_variant m.rroot)
                    | Some value ->
                        if has_suffix value "]" then
                          if not (has_prefix value "[") then try_roots rest
                          else
                            let decoded = decode_arbitrary_value (String.sub value 1 (String.length value - 2)) in
                            if (not (is_valid_arbitrary decoded)) || String.trim decoded = "" then None
                            else
                              Some
                                (make_variant ~vvalue:(Some { vvkind = Arbitrary; vvalue = decoded }) ~vmodifier ~vraw:input
                                   Functional_variant m.rroot)
                        else if has_suffix value ")" then
                          if not (has_prefix value "(") then try_roots rest
                          else
                            let decoded = decode_arbitrary_value (String.sub value 1 (String.length value - 2)) in
                            if (not (is_valid_arbitrary decoded)) || String.trim decoded = "" || not (has_prefix decoded "--") then None
                            else
                              Some
                                (make_variant ~vvalue:(Some { vvkind = Arbitrary; vvalue = "var(" ^ decoded ^ ")" }) ~vmodifier
                                   ~vraw:input Functional_variant m.rroot)
                        else if not (is_named_value value) then try_roots rest
                        else Some (make_variant ~vvalue:(Some { vvkind = Named; vvalue = value }) ~vmodifier ~vraw:input Functional_variant m.rroot)))
            | Some Compound_variant -> (
                match m.rvalue with
                | None -> None
                | Some inner_text ->
                    let inner_text, remaining =
                      if (m.rroot = "not" || m.rroot = "has" || m.rroot = "in") && modifier_text <> None then
                        (inner_text ^ "/" ^ Option.get modifier_text, None)
                      else (inner_text, modifier_text)
                    in
                    match ds.parse_variant inner_text with
                    | None -> None
                    | Some inner ->
                        if not (ds.compounds_with m.rroot inner) then None
                        else
                          let modifier =
                            match remaining with
                            | Some t -> ( match parse_modifier t with Some md -> Some (Some md) | None -> None)
                            | None -> Some None
                          in
                          (match modifier with
                          | None -> None
                          | Some vmodifier -> Some (make_variant ~vmodifier ~inner:(Some inner) ~vraw:input Compound_variant m.rroot)))
            | Some Arbitrary_variant -> try_roots rest)
      in
      try_roots (find_roots name ds.has_variant)
